### PR TITLE
add-janus-model

### PR DESCRIPTION
**Pull request by hsilva664**
# Nex fixes

- Adding JanusVisionModel as a proper wrapper
- Fixing multi-prompt image generation
- All tests have passed, but we still need to update the files in the HUB to reflect the new changes

### CODE REVIEW

File: src/transformers/models/janus/modular_janus.py [hunk 1]
```diff
@@ -1119,8 +1122,7 @@ def __init__(self, config: JanusConfig):
         super().__init__(config)
         self.config = config
         # This is necessary for backward compatibility, see SiglipModel initialization
-        tmp_vision_model = JanusVisionModel._from_config(config.vision_config)
-        self.vision_model = tmp_vision_model.vision_model
+        self.vision_model = JanusVisionModel._from_config(config.vision_config)
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py:885]: If we are creating the vision model then is there a need for the overwriting sdpa tests ig. IMO we should be able to do the below code change and this should fix set the attn_implementation.

`vision_model = JanusVisionModel._from_config(config)`

File: src/transformers/models/janus/modular_janus.py [hunk 2]
```diff
@@ -731,11 +732,13 @@ class JanusVisionModel(SiglipVisionModel):
 
     def __init__(self, config: JanusVisionConfig):
         super().__init__(config)
-
-        self.vision_model = JanusVisionTransformer(config)
-
+        del self.vision_model
+        self.vision_transformer = JanusVisionTransformer(config)
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py]: This one feels odd when considering repo consistency, lets instantiate the the vision model and since it's a wrapper around vision transformers so no harm in logits and keep this code design intact. Also vision transformers is not common vision backbone name we see, we generally keep it as vision_tower, vision_model etc.
```suggestion
        self.vision_model = JanusVisionModel(config)
```

File: src/transformers/models/janus/modular_janus.py [hunk 3]
```diff
@@ -731,11 +732,13 @@ class JanusVisionModel(SiglipVisionModel):
...
         self.post_init()
 
+    def forward(self, *args, **kwargs):
+        return self.vision_transformer(*args, **kwargs)
+
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py]: We can't use this as its too messy (don't know which argument is getting passed in code flow). Pass the standard args `pixel_values,return_dict,output_attention etc.. and remaining as kwargs`. Infact this seems to be a standard vsiion model module, so check can we use modular to import this module from another model

File: src/transformers/models/janus/modular_janus.py [hunk 4]
```diff
@@ -1390,7 +1398,7 @@ def prepare_inputs_for_generation(
                 batch_size, sequence_length = input_ids.shape
                 device = input_ids.device
 
-            attention_mask = self.model.language_model._prepare_4d_causal_attention_mask_with_cache_position(
+            attention_mask = self.model._prepare_4d_causal_attention_mask_with_cache_position(
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py]: Do we need this change? With this change we are adding an extra function.

File: src/transformers/models/janus/modular_janus.py [hunk 5]
```diff
@@ -1544,18 +1557,29 @@ def generate(
         decoder_hidden_states = () if (return_dict_in_generate and output_hidden_states) else None
         decoder_attentions = () if (return_dict_in_generate and output_attentions) else None
 
+
         for i in tqdm(range(num_image_tokens)):
-            # Fix me: What to do with attention mask when expanding and repeating input ids.
-            # Should we also modify the attention mask if passed?
-            outputs = self.model.language_model(
+            model_inputs = super().prepare_inputs_for_generation(
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py]: This is not the correct implementation, we already have a prepare_input_for_generation function defined above and there we have to do this super call. I have made the changes and will push it in sometime

File: src/transformers/models/janus/modular_janus.py [hunk 6]
```diff
@@ -1510,8 +1519,12 @@ def generate(
         num_image_tokens = self.model.vision_model.config.num_image_tokens
 
         input_tokens = input_ids.repeat(2, 1)  # Double batch size for conditional/unconditional logits
+        attention_mask = model_kwargs.pop("attention_mask", None) 
+        attention_mask = attention_mask.repeat(2, 1)
+
+        input_tokens[batch_size:, :].masked_fill_(input_tokens[batch_size:, :] != generation_config.bos_token_id,
+                                                  generation_config.pad_token_id)
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py]: Need some more clarification - I think in the original implementation they are masking every token except first and last which would be `beign_of_sentence_token` and `begin_of_image_token`(bos,boi). and here we are doing tokens !=bos which would also mask boi token

File: src/transformers/models/janus/modular_janus.py [hunk 7]
```diff
@@ -1572,6 +1596,9 @@ def generate(
             img_embeds = self.prepare_embeddings_for_image_generation(next_token)
             inputs_embeds = img_embeds.unsqueeze(dim=1)
 
+            # similar to GenerationMixin._sample, this is needed in prepare_inputs_for_generation
+            input_tokens = torch.cat([input_tokens, next_token[:, None]], dim=-1)
```
Comment (yaswanth19) [inline src/transformers/models/janus/modular_janus.py]: Can you clarify why this is required, AFAIK input embeds are used for first generation step and later on the next generated token is used along with cache. so we process next token and for previous context use the cache but here we are concating the input_tokens with next token, IMO we are passing the entire context everytime.

File: src/transformers/models/janus/modular_janus.py (hunk 2 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: I changed the name because otherwise the name of the variable would be `model.vision_model.vision_model` in our `state_dict`, which I think is not desired. Siglip does it like this, but, again, siglip uses that `tmp` construct that we've agreed to change (i.e., create a `VisionModel` and then passes its inner `vision_model` attribute and the parent gets deleted). What name do you want then?

File: src/transformers/models/janus/modular_janus.py (hunk 3 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: Will check once we decide on the name of the argument. Will either add the explicit arguments like you are suggesting or leave these lines out depending on your decision.

File: src/transformers/models/janus/modular_janus.py (hunk 2 above)
Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: Ohh yeah then we can keep `model = JanusVisionModel()`. Then we will have model.vision_model.model.vision_transformer. Please refer to `siglip2` codebase. The problem with siglip is they have used tmp for BC of previous checkpoints as it was integrated a few years ago.

File: src/transformers/models/janus/modular_janus.py (hunk 4 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: After the other changes, `generate` was trying to call (under the hood) `_prepare_4d_causal_attention_mask_with_cache_position` right from `JanusForConditionalGeneration` in one of the functions below and throwing a warning because it did not have it.

File: src/transformers/models/janus/modular_janus.py (hunk 1 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py:885]: `sdpa` parent tests---which you are suggesting reverting to---try to find `{"visual", "image_tower", "vision_tower", "vision_model"}` inside both `JanusModel` and `JanusForConditionalGeneration`. For the former it works OK, but it throws errors for the latter, because it should look for `obj.model.vision_model` in this case, not `obj.vision_model`. You can verify that the test fails by commenting the code from `test_sdpa_can_dispatch_composite_models` and calling `super().test_sdpa_can_dispatch_composite_models` instead.

File: src/transformers/models/janus/modular_janus.py (hunk 6 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: I missed that they were also leaving `boi` because of the -1, will add the changes (ps: the code works even without it). It is necessary to look for the positions instead of just assuming that the first position has `bos` bacause the first will not have `bos` in cases of left padding.

File: src/transformers/models/janus/modular_janus.py (hunk 5 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: Will look at what it does more closely and make the change, but wasn't the reason for the new `prepare_inputs_for_generation` something related to the image tokens for the cases where the user is passing an image (which I think it is not in this case)? Can you clarify why was it necessary to override?

File: src/transformers/models/janus/modular_janus.py (hunk 7 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: I added when I was testing my changes simply to be consistent with how the `GenerationMixin._sample` was doing it, and they do pass it to their `prepare_inputs_for_generation` function. Will double-check if it is indeed unnecessary in this function and remove it accordingly.

File: src/transformers/models/janus/modular_janus.py (hunk 6 above)
Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: I am really not sure, in ideal case scenario we should not mask bos and boi token according to their codebase but  how is it working even after masking boi for uncodition logits 😅 

> It is necessary to look for the positions instead of just assuming that the first position has bos bacause the first will not have bos in cases of left padding.

Just try to mask everything except boi and bos. If it's working fine then good else need to deep dive and find the optimal logic. This multi prompt case is not defined in their example nor they have anything related to training so we can implement this feature by our logic and just make it work

File: src/transformers/models/janus/modular_janus.py (hunk 5 above)
Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: For VLMs we have generally overwrite prepare_input_for_generation function to handle pixel values as it is not needed after first token is generated. I have made the change and pushed it. see if the function works for you and if not then we need to refactor there.

Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: For VLMs we have generally overwrite prepare_input_for_generation function to handle pixel values as it is not needed after first token is generated. I have made the change and pushed it. see if the function works for you and if not then we need to refactor there.

File: src/transformers/models/janus/modular_janus.py (hunk 2 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: Did not understand what you are proposing. How would we end with `model....model....`?

Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: model(janusmodel).vision_model(janus vison model).model(janus vision transformer0... from janusforconditionalgeneration perspective

Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: Just looked at siglip2. Their Siglip2Model is instantiated exactly like SiglipModel. WIll paste an excerpt below. This is the implementation we have been instructed to not use:
```python
        # First, initialize the text and vision models with proper attention implementation
        text_model = Siglip2TextModel._from_config(text_config)
        vision_model = Siglip2VisionModel._from_config(vision_config)

        # Second, get the text and vision submodules (for backward compatibility)
        self.text_model = text_model.text_model
        self.vision_model = vision_model.vision_model
```

Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: ouch missed this. Okay let's leave the sdpa case for now as we can always overwrite it based on our final model structure.

File: src/transformers/models/janus/modular_janus.py (hunk 6 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: I will make the change, I masked boi because I did not see it. Multi-prompt is not in their examples, but, from my inspections, HF designed things so that, by the use of position ids, the model will see the input like it would if a single image was passed. You can test that the version in this PR works.

File: src/transformers/models/janus/modular_janus.py (hunk 2 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: I haven't seen examples with `vision_model.model`. I can keep the name duplicated, as perhaps there is resistance to using `transformer` in the name. I simply used it because of the class name. Should I keep `vision_model.vision_model`? I might be able to avoid overwriting the forward pass like this.

Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: Once take a pull from my latest changes, I have changed the attention interface itself and the `test_dispatch_sdpa-composite` testcase is failing 😅 . Haven't deep dived into it to resolve but this one testcase is bugging a lot. So, first lets build the model structure and later override it.

I am a bit hesitant with hacky way we are using rn like del vision model is not right and even the fix using tmp variable is not optimal as in siglip it was used for BC and I don't see any newly merged models have this type of flow.

you can go through some vlm codebases (llava,idefics,paligemma,qwen etc) and try design the vision module similar to how its defined there. WDYT?

Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: PS: deleting is just a way to avoid the unraveling from including the attribute, they instruct us to do it like this in that modular-module tutorial. The `del ...` line does not appear in the modeling file. And having JanusVisionModel as a proper wrapper saves us from having to delete the tmp like they do.

Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: I will look

File: src/transformers/models/janus/modular_janus.py (hunk 7 above)
Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: Am glancing at the code now. As I have said `GenerationMixin._sample`---which we use for language model---updates input_ids like this and also, the function `GenerationMixin.prepare_inputs_for_generation`, which we use, even in your new branch, is also making heavy use of input ids. I am still checking there if it is OK to simply ignore this or pass an outdated input_ids every iteration. Passing None throws errors.

Comment (hsilva664) [reply src/transformers/models/janus/modular_janus.py]: It appears this can be changed to `input_tokens = next_token[:, None]`

Comment (yaswanth19) [reply src/transformers/models/janus/modular_janus.py]: > input_tokens = next_token[:, None]

Yup, thats what I was saying, we only need to pass new ids internally it should utilize cache for previous context.

**Comment by yaswanth19** [comment]
@hsilva664 Also can you check is the generate output consistent with that of original codebase. The forward pass is giving same outputs but the generated tokens were changing for formula example.

I did Janusforconditiongeneration(**inputs) and given the same input the output is same, so there is no issue in model code but something in generate might have been messed up although I didn't introduce any breaking change.
